proposals/propertyDependencies.md

```diff
@@ -0,0 +1,264 @@
+All examples have the same validation behavior as the examples above.
+
+This version uses an array of objects. Each object is a collection of the
+variables needed to express a property dependency. This doesn't fit the style of
+JSON Schema. There aren't any keywords remotely like this. It's also still too
+verbose. It's a little more intuitive than `if`/`then` and definitely less error
+prone.
+
+```json
+{
+  "propertyDependencies": [
+    {
+      "propertyName": "foo",
+      "propertySchema": { "const": "aaa" },
+      "apply": { "$ref": "#/$defs/foo-aaa" }
+    },
+    {
+      "propertyName": "foo",
+      "propertySchema": { "const": "bbb" },
+      "apply": { "$ref": "#/$defs/foo-bbb" }
+    }
+  ]
+}
+```
+
+A slight variation on that example is to make it a map of keyword to dependency
+object. It's still too verbose.
+
+```json
+{
+  "propertyDependencies": {
+    "foo": [
+      {
+        "propertySchema": { "const": "aaa" },
+        "apply": { "$ref": "#/$defs/foo-aaa" }
+      },
+      {
+        "propertySchema": { "const": "bbb" },
+        "apply": { "$ref": "#/$defs/foo-bbb" }
+      }
+    ]
+  }
+}
+```
+
+This one is a little more consistent with the JSON Schema style (poor keyword
+naming aside), but otherwise has all the same problems as the other examples.
+
+```json
+{
+  "allOf": [
+    {
+      "propertyDependencyName": "foo",
+      "propertyDependencySchema": { "const": "aaa" },
+      "propertyDependencyApply": { "$ref": "#/$defs/foo-aaa" }
+    },
+    {
+      "propertyDependencyName": "foo",
+      "propertyDependencySchema": { "const": "bbb" },
+      "propertyDependencyApply": { "$ref": "#/$defs/foo-bbb" }
+    }
+  ]
+}
+```
+
+This one is a variation of `if` that combines `if`, `properties`, and `required`
+to reduce boilerplate. It's also essentially a variation of the previous example
+with better names. This avoids to error proneness problem, but it's still too
+verbose.
+
+```json
+{
+  "allOf": [
+    {
+      "ifProperties": {
+        "foo": { "const": "aaa" }
+      },
+      "then": { "$ref": "#/$defs/foo-aaa" }
+    },
+    {
+      "ifProperties": {
+        "foo": { "const": "bbb" }
+      },
+      "then": { "$ref": "#/$defs/foo-aaa" }
+    }
+  ]
+}
+```
+
+All of the previous alternatives use a schema as the discriminator. This
+alternative is a little less powerful in that it can only match on exact values,
+but it successfully addresses the problems we're concerned about with the
+current approaches. The only issue with this alternative is that it's not as
+intuitive as the chosen solution.
+
+```json
+{
+  "propertyDepenencies": {
+    "foo": [
+      ["aaa", { "$ref": "#/$defs/foo-aaa" }],
+      ["bbb", { "$ref": "#/$defs/foo-bbb" }]
+    ]
+  }
+}
+```
+
+## Specification
+
+<!-- TODO: Update to where this actually ends up getting hosted. This is a guess
+for now. -->
+<https://json-schema.org/specification/extensions/propertyDependencies.html>
```

Comment:
But here you're linking to a published doc.  Seems like publishing this doc would help build a case for the feature's inclusion.

Reply:
We were talking about this as an ADR type of document. Usually that kind of thing has an internal audience, so I was expecting that this would probably just be on GitHub. I'm not opposed to thinking about putting it on the website for more exposure, but I wouldn't put it up in the form of a spec document like our main specs or extension specs. What I had in mind, was that part of the proposal making it to Stage-1 (or whatever we end up calling it) was adding a UJS entry describing the keyword. That would expose users to proposed keywords that they might want to use and we can link to the proposal and the spec for people who want to dig deeper.

Reply:
Publishing it leans into my previous point about a dev needing to convince their mgr/superior to use an experimental feature.  I think this content is useful.  Another reason I prefer the single doc approach.

Your concern before was that a dev wouldn't care about the "fluff" but would want to know about the stuff at the bottom and they shouldn't have to scroll all the way down there:

> the implementer needs to scroll past a bunch of stuff they don't care about in order to get to the information they need

They'll get a TOC that they can use to go straight to what they need.

Reply:
> Publishing it leans into my previous point about a dev needing to convince their mgr/superior to use an experimental feature.

Good point. That's a reason to consider having the audience be more than just internal. However, I'm not sure that necessarily means it should be on the website. The document is still public on GitHub.

> I think this content is useful for consumers of JSON Schema.

I think having something specifically for consumers and authors in UJS would be more useful because it's specifically focused for that audience.

> Another reason I prefer the single doc approach.

I'm not seeing the connection here. I think the separate documents approach makes sense in this case. The manager cares about what the trade-offs are. They can get that just from the proposal document. The consumer/author cares about what the feature does and how to use it. They can get that from just the UJS documentation. The implementer cares about what they need to build. They can get that from just the spec. Of course there's going to be some overlap and people wearing multiple hats, but I think the separation of concerns by audience makes a lot of sense.

> They'll get a TOC that they can use to go straight to what they need.

I take your point, but that doesn't address the other concerns I brought up. An extension could have a life outside of the proposal process. It might start off as an extension spec that was later proposed to be added to the spec. It might have been a proposal that was rejected, but still wants to live on as a third-party extension. Decoupling the spec from the proposal process allows for these cases.